product-classification/product_classification/InferenceEngine.py

```diff
@@ -0,0 +1,254 @@
+            )
+
+        combined["category_thresholded"] = combined.apply(
+            lambda x: x["predicted_category"]
+            if x["confidence"] >= x["threshold"]
+            else "under_threshold",
+            axis=1,
+        )
+
+        return combined[list(classification_df.columns) + ["category_thresholded", "threshold"]]
+
+    def probas_to_ProductClassifications(self, probas: pd.DataFrame) -> pd.DataFrame:
+        """
+        This function is used to transform predicted probabilities into a DataFrame with complete
+        ProductClassification objects.
+
+        Args:
+            probas (pd.Dataframe): Dataframe with predicted probabilities.
+
+        Returns:
+            pd.DataFrame: pd.DataFrame of ProductClassification objects.
+        """
+
+        probas.columns = le.inverse_transform(probas.columns)
+        predicted_category = [probas.columns[np.argmax(p)] for p in probas.values]
+        confidence = [np.max(p) for p in probas.values]
+
+        result = pd.DataFrame(
+            {
+                "id": probas.index,
+                "ml_model_name": self.name,
+                "predicted_category": predicted_category,
+                "confidence": confidence,
+                "all_predicted_probabilities": probas.to_dict(orient="records"),
+            }
+        )
+
+        return result
+
+    def run_pipeline(self, request_data: str, apply_thresholds: bool = False) -> pd.DataFrame:
+        """The standard pipeline to run for evaluating the request data and performing inference.
+
+        Args:
+        request_data (json): data which was sent along the POST request. Should include a
+        pd.DataFrame that was transformed into serialized json format.
+        data_format (str): format to use for evaluating the request data. Either 'products' or
+        'classifications'.
+
+        :return:
+            A flask Response containing the predictions.
+        """
+        df = self.eval_request(request_data)
+        pred_probs = self.predict_proba(df)
+
+        classification_df = self.probas_to_ProductClassifications(pred_probs)
+        if apply_thresholds:
+            classification_df = self.apply_shop_thresholds(classification_df, df)
+
+        return classification_df
+
+    @staticmethod
+    def eval_request(request_data: str, data_format: str = "products") -> Optional[pd.DataFrame]:
+        """
+        This function is used to evaluate the data of the request and checks whether all necessary
+        columns/features are part of the request data.
+
+        Args: request_data (str): data which was sent along the POST request with a string in
+        JSON format. data_format (str): format to use for evaluating the request data. Either
+        'products' or 'classifications'.
+
+
+        Returns:
+            Optional(pd.DataFrame): pd.DataFrame of the request data.
+        """
+
+        df = pd.read_json(request_data)
+
+        if data_format == "products":
```

Comment:
Maybe add a comment line regarding the different data formats and scenarios, or add that in the docs above